pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+            >>> sc = b'ABCDEFGHIGKLMNOPQRSTUVXYZ'
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> encoders.i386.ascii_shellcode.encode._get_subtractions(sc, vocab)
+            b'-(!!!-~NNNP-!=;:-f~~~-~~~~P-!!!!-edee-~~~~P-!!!!-eddd-~~~~P-!!!!-egdd-~~~~P-!!!!-eadd-~~~~P-!!!!-eddd-~~~~P'
+        """
+        int_size = context.bits // 8
+        result = bytes()
+        last = b'\x00'*int_size
+        # Group the shellcode into bytes of stack cell size, pad with NOPs
+        # if the shellcode does not divide into stack cell size and reverse.
+        # The shellcode will be reversed again back to it's original order once
+        # it's pushed onto the stack
+        if six.PY3:
+            sc = tuple(group(int_size, shellcode, 0x90))[::-1]
+        elif six.PY2:
+            sc = []
+            for byte in group(int_size, shellcode, b'\x90'):
+                sc.append(''.join(byte))
+            sc = sc[::-1]
+        # Pack the shellcode to a sub/push sequence
+        for x in map(bytes, sc):
+            for subtraction in self._calc_subtractions(last, x, vocab):
+                if six.PY2:
+                    subtraction = str(subtraction)
+                # sub eax, `subtraction`
+                result += struct.pack('=c{}s'.format(int_size), b'-', subtraction)
+            last = x
+            result += b'P'  # push eax
+        return result
+
+
+    @LocalContext
+    def _calc_subtractions(self, last, target, vocab, max_subs = 4):
+        r""" Given `target` and `last`, return a list of integers that when
+         subtracted from `last` will equal `target` while only constructing
+         integers from bytes in `vocab`
+
+        int_size is take from the context (context.bits / 8)
+
+        Args:
+            last (bytes): Current value of eax
+            target (bytes): Desired value of eax
+            vocab (bytes): Allowed characters
+            max_subs (int): Maximum subtraction attempts
+
+        Raises:
+            ArithmeticError: If a sequence of subtractions could not be found
+
+        Returns:
+            List[bytearray]: List of numbers that would need to be subtracted
+            from `last` to get to `target`
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x10'*4, b'\x11'*4, vocab))
+            [bytearray(b'!!!!'), bytearray(b'`___'), bytearray(b'~~~~')]
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x11\x12\x13\x14', b'\x15\x16\x17\x18', vocab))
+            [bytearray(b'~}}}'), bytearray(b'~~~~')]
+        """
+        int_size = context.bits // 8
+        subtractions = [bytearray(b'\x00'*int_size)]

```

Comment:
In case you didn't know
```suggestion
        subtractions = [bytearray(int_size)]
```